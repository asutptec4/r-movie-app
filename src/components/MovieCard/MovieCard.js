import PropTypes from 'prop-types';
import React from 'react';

import { DELETE_ACTION, EDIT_ACTION } from '../../constant';
import { movie } from '../../types/movie';
import MoviePoster from '../MoviePoster/MoviePoster';
import SingleSelectDropdown from '../SingleSelectDropdown/SingleSelectDropdown';
import './MovieCard.scss';

const editOptions = [
  { id: EDIT_ACTION, name: 'Edit' },
  { id: DELETE_ACTION, name: 'Delete' },
];

const Button = () => {
  return (
    <div className="movie-options-button">
      <div>...</div>
    </div>
  );
};

const MovieCard = ({ movie, handleCardAction, handleCardClick }) => {
  return (
    <div className="movie-card">
      <div className="movie-image" onClick={(e) => handleCardClick(movie)}>
        <MoviePoster imageUrl={movie.poster} />
      </div>
      <div className="movie-desc">
        <span className="title">{movie.title}</span>
        <span className="genre">{movie.genres ? movie.genres.join(' ') : ''}</span>
        <span className="year">{movie.releaseYear}</span>
      </div>
      <div className="button">
        <SingleSelectDropdown
          options={editOptions}
          customButton={<Button />}
          onOptionChange={(o) => handleCardAction(o?.id, movie)}
        />
      </div>
    </div>
  );
};

MovieCard.propTypes = {
  movie: movie,
  handleCardAction: PropTypes.func.isRequired,
  handleCardClick: PropTypes.func.isRequired,
};

export default MovieCard;
