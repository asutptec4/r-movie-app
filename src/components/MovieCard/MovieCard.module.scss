@import '../../scss/index.scss';

$img-w: 360px;

.movieCard {
  margin-bottom: 30px;
  margin: 1em auto;
  position: relative;

  .movieImage {
    margin: 0 auto;
    width: $img-w;
    height: calc(4 / 3 * #{$img-w});
    cursor: pointer;
  }

  .button {
    position: absolute;
    top: 15px;
    left: calc(0.75 * #{$img-w});
    display: none;

    .movieOptionsButton {
      background-color: $bg-color;
      border-radius: 50%;
      width: 40px;
      height: 40px;
      margin-left: 40px;

      div {
        transform: rotate(90deg) translate(20px, -4px);
        font-weight: bold;
      }
    }
  }

  .movieDesc {
    width: $img-w;
    padding: 0.5em;
    display: grid;
    grid-template-columns: auto auto;
    grid-template-rows: auto auto;

    .title {
      font-size: $fs-normal;
    }

    .genre {
      font-size: $fs-small;
      grid-row: 2;
    }

    .year {
      grid-column: 3;
      justify-self: end;
      padding: 0.25em 0.5em;
      border: 1px solid;
      border-radius: 3px;
    }
  }
}

.movieCard:hover .button {
  display: block;
}
