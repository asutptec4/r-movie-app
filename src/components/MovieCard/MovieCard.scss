@import '../../scss/index.scss';

$img-w: 360px;

.movie-card {
  margin-bottom: 30px;
  margin: 1em auto;

  .movie-image {
    padding: 2px;
    width: $img-w;
    height: calc(4/3 * #{$img-w});
  }

  .movie-desc {
    width: $img-w;
    padding: 5px;
    display: grid;
    grid-template-columns: auto auto;
    grid-template-rows: auto auto;

    .title {
      font-size: $fs-normal;
    }

    .genre {
      font-size: $fs-small;
      grid-row: 2;
    }

    .year {
      grid-column: 3;
      justify-self: end;
      padding: 5px 10px;
      border: 1px solid;
      border-radius: 3px;
    }
  }
}
