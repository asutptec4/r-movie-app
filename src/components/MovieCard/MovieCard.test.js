import { fireEvent, render } from '@testing-library/react';
import React from 'react';

import MovieCard from './MovieCard';

jest.mock('../MoviePoster/MoviePoster', () => 'MoviePosterComponent');

describe('<MovieCard />', () => {
  const mockCardActionHandler = jest.fn();
  const mockCardClickHandler = jest.fn();
  const defaultProps = {
    movie: {
      genres: ['Science Fiction', 'Action', 'Adventure'],
      id: 424785,
      overview: 'Plot unknown.',
      poster: 'https://image.tmdb.org/t/p/w500/432BowXw7a4fWXSONxBaFKqvW4f.jpg',
      rating: '8.0',
      releaseDate: '2019-06-26',
      releaseYear: '2019',
      runtime: 120,
      subtitle: '',
      title: 'Transformers 7',
    },
    handleCardAction: mockCardActionHandler,
    handleCardClick: mockCardClickHandler,
  };

  test('should render movie prop', () => {
    const { container } = render(<MovieCard {...defaultProps} />);
    expect(container).toMatchSnapshot();
  });

  test('should invoke handleCardClick on image click', () => {
    const { container } = render(<MovieCard {...defaultProps} />);
    fireEvent.click(container.firstChild.firstChild);
    expect(mockCardClickHandler).toHaveBeenCalledWith(defaultProps.movie);
  });

  test('should print empty string if genres are not provided', () => {
    defaultProps.movie.genres = undefined;
    const { container } = render(<MovieCard {...defaultProps} />);
    const genre = container.getElementsByClassName('genre')[0];
    expect(genre.innerHTML).toEqual('');
  });
});
