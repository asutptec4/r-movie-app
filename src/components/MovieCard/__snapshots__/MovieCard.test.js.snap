// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MovieCard /> should render movie prop 1`] = `
<div>
  <div
    class="movie-card"
  >
    <div
      class="movie-image"
    >
      <moviepostercomponent
        imageurl="https://image.tmdb.org/t/p/w500/432BowXw7a4fWXSONxBaFKqvW4f.jpg"
      />
    </div>
    <div
      class="movie-desc"
    >
      <span
        class="title"
      >
        Transformers 7
      </span>
      <span
        class="genre"
      >
        Science Fiction Action Adventure
      </span>
      <span
        class="year"
      >
        2019
      </span>
    </div>
    <div
      class="button"
    >
      <div
        class="single-select-dropdown-container"
      >
        <div
          class="single-select-dropdown-button"
        >
          <div
            class="movie-options-button"
          >
            <div>
              ...
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
